### PR TITLE
Preferences: choose dark or light icons depending on OS theme

### PR DESCRIPTION
**Pull request by ronso0**
... because previously dark icons were hard to recognise with dark themes.

This is a rather simple approach that checks whether the palette's primary text color is dim or not and chooses the respective icon set. The two sets of dark and light icons include all icons in the side bar and those in the controller preset drop-down menu.
I removed the gradients and use plain color fill colors now. Icon folders are much smaller now because of 'optimized SVGs'.

To allow adding the icon set switch I refactored DlgPreferences to add the individual Pref page widgets and tree icons in a much more compact way in `addPageWidget()` insetad of separate slots.

image

### CODE REVIEW

**Comment by ronso0** [comment]
In conjunction with #3464 the preferences should now be usable with all sorts of themes.

Menubar theming per skin is on its way... :)

**Comment by ronso0** [comment]
@Holzhaus Can you check how that looks like with i3wm?

**Comment by ronso0** [comment]
@Be-ing Did you try this? Any comments on the code?

**Comment by Holzhaus** [comment]
> @Holzhaus Can you check how that looks like with i3wm?

Sorry, I switched to pop-shell recently.

File: src/controllers/dlgprefcontroller.cpp [hunk 1]
```diff
@@ -311,15 +311,21 @@ void DlgPrefController::enumeratePresets(const QString& selectedPresetPath) {
 
     // qDebug() << "Enumerating presets for controller" << m_pController->getName();
 
+    // Check the text color of the palette for whether to use dark or light icons
+    QString iconsPath = ":/images/preferences/dark/";
+    if (!Color::isDimColor(palette().text().color())) {
+        iconsPath = ":/images/preferences/light/";
+    }
```
Comment (Holzhaus) [inline src/controllers/dlgprefcontroller.cpp:318]: According to the docs we could also use `QDir` to access the resource system (didn't test it though). If that works we can get rid of the string concatenation:

```suggestion
    QString iconsPath = QDir(":/images/preferences/dark/") ;
    if (!Color::isDimColor(palette().text().color())) {
        iconsPath.setPath(":/images/preferences/light/");
    }
```

Then you can build paths like this which looks nicer to me:
```cpp
const auto noPresetIcon = QIcon(iconsPath.filePath("ic_none.svg"));
```

File: src/controllers/dlgprefcontrollers.cpp [hunk 2]
```diff
@@ -41,26 +43,26 @@ void DlgPrefControllers::slotOpenLocalFile(const QString& file) {
 }
 
 void DlgPrefControllers::slotUpdate() {
-    for (DlgPrefController* pControllerWindows : qAsConst(m_controllerWindows)) {
-        pControllerWindows->slotUpdate();
+    for (DlgPrefController* controllerDlg : qAsConst(m_controllerPages)) {
+        controllerDlg->slotUpdate();
```
Comment (Holzhaus) [inline src/controllers/dlgprefcontrollers.cpp:47]:
```suggestion
    for (DlgPrefController* pControllerDlg : qAsConst(m_controllerPages)) {
        pControllerDlg->slotUpdate();
```

and below...

File: src/controllers/dlgprefcontrollers.cpp [hunk 3]
```diff
@@ -71,12 +73,12 @@ QUrl DlgPrefControllers::helpUrl() const {
 bool DlgPrefControllers::handleTreeItemClick(QTreeWidgetItem* clickedItem) {
     int controllerIndex = m_controllerTreeItems.indexOf(clickedItem);
     if (controllerIndex >= 0) {
-        DlgPrefController* controllerWidget = m_controllerWindows.value(controllerIndex);
-        if (controllerWidget) {
-            m_pDlgPreferences->switchToPage(controllerWidget);
+        DlgPrefController* controllerDlg = m_controllerPages.value(controllerIndex);
+        if (controllerDlg) {
+            m_pDlgPreferences->switchToPage(controllerDlg);
```
Comment (Holzhaus) [inline src/controllers/dlgprefcontrollers.cpp:78]:
```suggestion
        DlgPrefController* pControllerDlg = m_controllerPages.value(controllerIndex);
        if (pControllerDlg) {
            m_pDlgPreferences->switchToPage(pControllerDlg);
```

File: src/controllers/dlgprefcontrollers.cpp [hunk 4]
```diff
@@ -91,16 +93,16 @@ void DlgPrefControllers::rescanControllers() {
 }
 
 void DlgPrefControllers::destroyControllerWidgets() {
-    while (!m_controllerWindows.isEmpty()) {
-        DlgPrefController* controllerDlg = m_controllerWindows.takeLast();
+    while (!m_controllerPages.isEmpty()) {
+        DlgPrefController* controllerDlg = m_controllerPages.takeLast();
         m_pDlgPreferences->removePageWidget(controllerDlg);
         delete controllerDlg;
```
Comment (Holzhaus) [inline src/controllers/dlgprefcontrollers.cpp:99]:
```suggestion
        DlgPrefController* pControllerDlg = m_controllerPages.takeLast();
        m_pDlgPreferences->removePageWidget(pControllerDlg);
        delete pControllerDlg;
```

File: src/controllers/dlgprefcontrollers.cpp [hunk 5]
```diff
@@ -91,16 +93,16 @@ void DlgPrefControllers::rescanControllers() {
...
     }
 
     m_controllerTreeItems.clear();
-    while(m_pControllerTreeItem->childCount() > 0) {
-        QTreeWidgetItem* controllerWindowLink = m_pControllerTreeItem->takeChild(0);
-        delete controllerWindowLink;
+    while (m_pControllersRootItem->childCount() > 0) {
+        QTreeWidgetItem* controllerTreeItem = m_pControllersRootItem->takeChild(0);
+        delete controllerTreeItem;
```
Comment (Holzhaus) [inline src/controllers/dlgprefcontrollers.cpp:105]:
```suggestion
        QTreeWidgetItem* pControllerTreeItem = m_pControllersRootItem->takeChild(0);
        delete pControllerTreeItem;
```

File: src/controllers/dlgprefcontrollers.cpp [hunk 6]
```diff
@@ -123,49 +132,44 @@ void DlgPrefControllers::setupControllerWidgets() {
...
-    // If no controllers are available, show the "No controllers available"
-    // message.
-    txtNoControllersAvailable->setVisible(controllerList.empty());
 }
 
-void DlgPrefControllers::slotHighlightDevice(DlgPrefController* dialog, bool enabled) {
-    int dialogIndex = m_controllerWindows.indexOf(dialog);
-    if (dialogIndex < 0) {
+void DlgPrefControllers::slotHighlightDevice(DlgPrefController* controllerDlg, bool enabled) {
+    int controllerPageIndex = m_controllerPages.indexOf(controllerDlg);
```
Comment (Holzhaus) [inline src/controllers/dlgprefcontrollers.cpp:161]:
```suggestion
void DlgPrefControllers::slotHighlightDevice(DlgPrefController* pControllerDlg, bool enabled) {
    const int controllerPageIndex = m_controllerPages.indexOf(pControllerDlg);
```

and also add the `p` prefix in the code below both for controllerDlg and controllerTreeItem...

File: src/preferences/dialog/dlgpreferences.cpp [hunk 7]
```diff
@@ -73,80 +74,180 @@ DlgPreferences::DlgPreferences(MixxxMainWindow * mixxx, SkinLoader* pSkinLoader,
...
-    // Construct widgets for use in tabs.
+    // Construct widgets for use in tabs:
+    // * check the text color of the palette for whether to use dark or light icons
+    // * create the tree view button first (important for 'Controllers' page!)
+    // * instantiate preferences page
+    QString iconsColor = "dark/";
+    if (!Color::isDimColor(palette().text().color())) {
+        iconsColor = QStringLiteral("light/");
+    }
+    m_iconsPath = ":/images/preferences/" + iconsColor;
```
Comment (Holzhaus) [inline src/preferences/dialog/dlgpreferences.cpp:89]: QDir?

File: src/preferences/dialog/dlgpreferences.cpp [hunk 8]
```diff
@@ -426,10 +388,11 @@ void DlgPreferences::onShow() {
     // Update geometry with last values
 #ifdef __WINDOWS__
     resize(m_geometry[2].toInt(), m_geometry[3].toInt());
-#else
-    // On linux, when the window is opened for the first time by the window manager, QT does not have
-    // information about the frame size so the offset is zero. As such, the first time it opens the window
-    // does not include the offset, so it is moved from the last position it had.
+#else /* __WINDOWS__ */
+    // On linux, when the window is opened for the first time by the window manager,
+    // QT does not have information about the frame size so the offset is zero.
```
Comment (Holzhaus) [inline src/preferences/dialog/dlgpreferences.cpp:393]:
```suggestion
    // Qt does not have information about the frame size so the offset is zero.
```

**Review by Holzhaus — COMMENTED** [review]
Thanks. Works fine on pop-shell/GNOME.

Looks like some unrelated mass reformattings slipped in. Please either remove them or fix the missing `p` prefix for all pointer types.

**Comment by ronso0** [comment]
> Thanks. Works fine on pop-shell/GNOME.
> 
> Looks like some unrelated mass reformattings slipped in. Please either remove them or fix the missing `p` prefix for all pointer types.

okay, I'll fix that after adding the QDir suggestion.
I'm more interested in if fd89d4742aa412a33842dba777a3f36ef56a4f90 is okay, then I'll fix the remaining issues in the other commits.

File: src/controllers/dlgprefcontroller.cpp (hunk 1 above)
Comment (ronso0) [reply src/controllers/dlgprefcontroller.cpp:318]: I'd do it like this:
``` c++
    QString iconsPath = ":/images/preferences/dark/";
    if (!Color::isDimColor(palette().text().color())) {
        iconsPath = ":/images/preferences/light/";
    }
    m_iconsPath = iconsPath;
```
and
``` c++
    treeItem->setIcon(0, QIcon(m_iconsPath.filePath(iconFile)));
```
?
Builds fine and works.
I suspect QDir::filePath() does the same string concatenation interanlly btw.

**Comment by ronso0** [comment]
I fixed the pointer prefixes and added the QDir suggestion.
Rebased onto fd89d4742aa412a33842dba777a3f36ef56a4f90

File: src/preferences/dialog/dlgpreferences.cpp (hunk 7 above)
Comment (ronso0) [reply src/preferences/dialog/dlgpreferences.cpp]: jep, done.

File: src/controllers/dlgprefcontrollers.cpp (hunk 6 above)
Comment (ronso0) [reply src/controllers/dlgprefcontrollers.cpp]: Done https://github.com/mixxxdj/mixxx/pull/3484/commits/f9742c598320341a2e5d96fe3c972c691213f7d1

File: src/controllers/dlgprefcontroller.cpp [hunk 9]
```diff
@@ -311,15 +312,21 @@ void DlgPrefController::enumeratePresets(const QString& selectedPresetPath) {
 
     // qDebug() << "Enumerating presets for controller" << m_pController->getName();
 
+    // Check the text color of the palette for whether to use dark or light icons
+    QDir iconsPath = QString(":/images/preferences/dark/");
+    if (!Color::isDimColor(palette().text().color())) {
+        iconsPath = QString(":/images/preferences/light/");
```
Comment (Holzhaus) [inline src/controllers/dlgprefcontroller.cpp:318]:
```
home/jan/Projects/mixxx/src/controllers/dlgprefcontroller.cpp: In member function ‘void DlgPrefController::enumeratePresets(const QString&)’:
/home/jan/Projects/mixxx/src/controllers/dlgprefcontroller.cpp:318:58: warning: ‘QDir& QDir::operator=(const QString&)’ is deprecated: Use QDir::setPath() instead [-Wdeprecated-declarations]
  318 |         iconsPath = QString(":/images/preferences/light/");
      |                                                          ^
In file included from /usr/include/qt/QtCore/QDir:1,
                 from /home/jan/Projects/mixxx/src/controllers/controllerpreset.h:4,
                 from /home/jan/Projects/mixxx/src/controllers/controllermappingtablemodel.h:10,
                 from /home/jan/Projects/mixxx/src/controllers/controllerinputmappingtablemodel.h:8,
                 from /home/jan/Projects/mixxx/src/controllers/dlgprefcontroller.h:6,
                 from /home/jan/Projects/mixxx/src/controllers/dlgprefcontroller.cpp:1:
/usr/include/qt/QtCore/qdir.h:110:11: note: declared here
  110 |     QDir &operator=(const QString &path);
      |           ^~~~~~~~
```

File: src/preferences/dialog/dlgpreferences.cpp [hunk 10]
```diff
@@ -73,80 +77,180 @@ DlgPreferences::DlgPreferences(MixxxMainWindow * mixxx, SkinLoader* pSkinLoader,
...
-    addPageWidget(m_soundPage);
+    // Construct widgets for use in tabs:
+    // * check the text color of the palette for whether to use dark or light icons
+    // * create the tree view button first (important for 'Controllers' page!)
+    // * instantiate preferences page
+    QString iconsPath = ":/images/preferences/dark/";
+    if (!Color::isDimColor(palette().text().color())) {
+        iconsPath = ":/images/preferences/light/";
+    }
+    m_iconsPath = iconsPath;
```
Comment (Holzhaus) [inline src/preferences/dialog/dlgpreferences.cpp:92]:
```
home/jan/Projects/mixxx/src/preferences/dialog/dlgpreferences.cpp: In constructor ‘DlgPreferences::DlgPreferences(MixxxMainWindow*, SkinLoader*, SoundManager*, PlayerManager*, ControllerManager*, VinylControlManager*, LV2Backend*, EffectsManager*, SettingsManager*, Library*)’:
/home/jan/Projects/mixxx/src/preferences/dialog/dlgpreferences.cpp:92:19: warning: ‘QDir& QDir::operator=(const QString&)’ is deprecated: Use QDir::setPath() instead [-Wdeprecated-declarations]
   92 |     m_iconsPath = iconsPath;
      |                   ^~~~~~~~~
In file included from /usr/include/qt/QtCore/QDir:1,
                 from /home/jan/Projects/mixxx/src/preferences/dialog/dlgpreferences.h:4,
                 from /home/jan/Projects/mixxx/src/preferences/dialog/dlgpreferences.cpp:1:
/usr/include/qt/QtCore/qdir.h:110:11: note: declared here
  110 |     QDir &operator=(const QString &path);
      |           ^~~~~~~~
```

**Review by Holzhaus — CHANGES_REQUESTED** [review]

File: src/controllers/dlgprefcontroller.cpp (hunk 9 above)
Comment (ronso0) [reply src/controllers/dlgprefcontroller.cpp:318]: Thank you. Locally I still build with Qt5.12.8 (version in Ubuntu 20.04) and with that I don't see deprecation warnings.
Changed to what you literally suggested above.

File: src/preferences/dialog/dlgpreferences.cpp (hunk 10 above)
Comment (ronso0) [reply src/preferences/dialog/dlgpreferences.cpp]: Fixed, too.

**Comment by ronso0** [comment]
QDir issues fixed, ready again.

File: src/controllers/dlgprefcontroller.cpp [hunk 11]
```diff
@@ -311,15 +312,21 @@ void DlgPrefController::enumeratePresets(const QString& selectedPresetPath) {
 
     // qDebug() << "Enumerating presets for controller" << m_pController->getName();
 
+    // Check the text color of the palette for whether to use dark or light icons
+    QDir iconsPath = QString(":/images/preferences/dark/");
```
Comment (uklotzde) [inline src/controllers/dlgprefcontroller.cpp:316]: Use a default constructed QDir and the setPath() in both symmetric if/else arms.

**Review by uklotzde — COMMENTED** [review]
Just some minor comments.

There are also many opportunities for wrapping string literals into QStringLiteral(). Qt legacy burden, unfortunately.

File: src/controllers/dlgprefcontroller.cpp (hunk 11 above)
Comment (ronso0) [reply src/controllers/dlgprefcontroller.cpp:316]: Okido, makes sense. If that's all you prefer to be changed I'd ammend those changes to 1b165bb752ab6cff3b986a8da169126d2e0e8527 and force-push

**Comment by ronso0** [comment]
> There are also many opportunities for wrapping string literals into QStringLiteral()

@uklotzde If you refer to the preset getter functions I could change that, but I'd rather do that in #3464

File: src/controllers/dlgprefcontroller.cpp (hunk 11 above)
Comment (ronso0) [reply src/controllers/dlgprefcontroller.cpp:316]: just drop a line here (no reaction emoji, I probably don't notice those)

Comment (ronso0) [reply src/controllers/dlgprefcontroller.cpp:316]: _ping_ @uklotzde

**Comment by uklotzde** [comment]
This is a already huge PR. No need for more refactoring here. I agree to defer those non-mandatory changes until later.

**Comment by ronso0** [comment]
so all LGTY?

**Comment by ronso0** [comment]
will fix the conflicts now

**Comment by ronso0** [comment]
> will fix the conflicts now

Done, CI all green

**Comment by ronso0** [comment]
oha! I tried merging 2.3 into main locally and that is fun already right now. Some not so trivial conflicts in dlgpreferences.cpp 
@uklotzde I guess you have a git rerere set for this.

After this PR merged some addPageWidget() optimizations will clash..
I volunteer for merging 2.3 > main then, looks feasible.

**Comment by ronso0** [comment]
reset to Draft. there's one !fixup commit I missed to squash. I'll do this when you consider this ready to be merged.

**Comment by ronso0** [comment]
> reset to Draft. there's one !fixup commit I missed to squash. I'll do this when you consider this ready to be merged.

_ping_ @uklotzde If this is fine now please let me know.
I'd then squash in the fixup commit and take care of merging 2.3 into main.

**Comment by ronso0** [comment]
> > reset to Draft. there's one !fixup commit I missed to squash. I'll do this when you consider this ready to be merged.
> 
> _ping_ @uklotzde If this is fine now please let me know.
> I'd then squash in fixup! [5f88ac2](https://github.com/mixxxdj/mixxx/commit/5f88ac201d63aeb296e2b36ff8e7c48119fe834d) and take care of merging 2.3 into main.

@Holzhaus Final LGTM?
I'd like to take care of this and fix the main regression asap. Won't have too much time soon.

**Comment by Holzhaus** [comment]
I didn't review this again because it's still marked as draft. Should I review?

**Comment by ronso0** [comment]
Yes please.
I marked it as draft to prevent merging before I squashed !fixup! 5f88ac2

Per commit this is easy to review, starting with what you already reviewed.

**Comment by Holzhaus** [comment]
```
Warning [Main]: Cannot open file ':/images/preferences/light/broadcast.svg', because: No such file or directory
Warning [Main]: Cannot open file ':/images/preferences/light/broadcast.svg', because: No such file or directory
```

File: src/preferences/dialog/dlgpreferences.cpp (hunk 8 above)
Comment (Holzhaus) [reply src/preferences/dialog/dlgpreferences.cpp:396]: ping

**Review by Holzhaus — COMMENTED** [review]
Code look good, thank you. When the missing icon error on start is fixed, we can merge IMHO.

**Comment by ronso0** [comment]
`Warning [Main]: Cannot open file ':/images/preferences/light/broadcast.svg', because: No such`

fixed.

**Comment by ronso0** [comment]
> Code look good, thank you. When the missing icon error on start is fixed, we can merge IMHO.

Great, thanks for being quick!

Seems like a created a small mess with those forgotten !fixup commits and merging 2.3 later on.
I'll re-apply my commits without 2.3 now so this is in a good shape finally.

**Comment by ronso0** [comment]
Puuh, done.
If CI is all green please merge and I'll take care of 2.3 > main

**Comment by ronso0** [comment]
@Holzhaus 
CI is happy

**Comment by ronso0** [comment]
Thanks. merging 2.3 to main now...huu fun

**Pull request merged**